lib/beaker-hostgenerator/generator.rb

```diff
@@ -1,23 +1,28 @@
 require 'beaker-hostgenerator/data'
-require 'beaker-hostgenerator/exceptions'
 require 'beaker-hostgenerator/roles'
 require 'beaker-hostgenerator/hypervisor'
+require 'beaker-hostgenerator/parser'
 
 require 'yaml'
 
 module BeakerHostGenerator
   class Generator
     include BeakerHostGenerator::Data
-    include BeakerHostGenerator::Exceptions
+    include BeakerHostGenerator::Parser
+    include BeakerHostGenerator::Roles
 
     # Main host generation entry point, returns a YAML map as a string for the
     # given host specification and optional configuration.
     #
-    # @param [Array<String>] tokens The hosts specification input split on '-'.
-    #                        For example `["centos6", "64m", "redhat7", "64a"]`.
-    # @param [Hash] options General configuration options, such as optional
-    #               parameters provided on the command line.
-    def generate(tokens, options)
+    # @param layout [String] The raw hosts specification user input.
+    #                        For example `"centos6-64m-redhat7-64a"`.
+    # @param options [Hash] Global, optional configuration such as the default
+    #                       hypervisor or OS info version.
+    #
+    # @returns [String] A complete YAML map as a string defining the HOSTS and
+    #                   CONFIG sections as required by Beaker.
+    def generate(layout, options)
```

Comment:
Like I said in my other comment, I tried many times to decompose this cleanly, but I didn't like any of them.
There's too many local variables weaved through this method, and I believe it could be rearranged to allow for better decomposition, but I didn't want to do too much all at once.

I did extract all the role code into its own method below, so that helps a bit.